src/main/lrsql/util/oidc.clj

```diff
@@ -1,19 +1,187 @@
+  [{:keys [jwks-uri] :as config}]
+  (try
+    (if-let [jwks-uri (or jwks-uri
+                          (some-> config
+                                  get-configuration
+                                  (get "jwks_uri")))]
+      (let [keyset-cache (atom (jwt/get-keyset jwks-uri))]
+        [;; Decode/Unsign tokens
+         (oidc-i/decode-interceptor
+          (fn [_]
+            (fn [kid]
+              (get @keyset-cache kid
+                   ;; If kid is not found in the keyset, attempt refresh and try
+                   ;; again
+                   (get (reset! keyset-cache (jwt/get-keyset jwks-uri))
+                        kid))))
+          :required? false)
+         ;; This is a vector in case we need additional interceptors. At present
+         ;; we do not.
+         ])
+      ;; If no config, don't return any
+      [])
+    (catch Exception ex
+      (throw
+       (ex-info
+        "OIDC Initialization Failure"
+        {:type        ::init-failure
+         :oidc-config (select-config config)}
+        ex)))))
+
+(s/fdef parse-scope-claim
+  :args (s/cat :scope-str string?)
+  :ret (s/every ::auth/scope))
+
+(defn parse-scope-claim
+  [scope-str]
+  (keep auth/scope-str->kw
+        (cs/split scope-str #"\s")))
+
+(s/fdef token-auth-identity
+  :args (s/cat :ctx map?
+               :authority-fn fn?)
+  :ret (s/nilable
+        ::lrs-auth/identity))
+
+(defn token-auth-identity
+  "For the given context, return a valid auth identity from token claims. If no
+  claims are present, return nil"
+  [ctx
+   authority-fn]
+  (when-let [token (:com.yetanalytics.pedestal-oidc/token ctx)]
+    (let [{:keys [scope]
+           :as   claims} (get-in ctx
+                               [:request
+                                :com.yetanalytics.pedestal-oidc/claims])]
+      {:result
+       (if-let [scopes (some-> scope
+                               parse-scope-claim
+                               not-empty
+                               (->> (into #{})))]
+         (try
+           {:scopes scopes
+            :prefix ""
+            :auth   {:token token}
+            :agent  (authority-fn claims)}
+           (catch clojure.lang.ExceptionInfo ex
+             (let [{ex-type :type} (ex-data ex)]
+               (case ex-type
+                 ::authority/unknown-variable
+                 (do
+                   (log/warn "Could not render authority for OIDC claims")
+                   :com.yetanalytics.lrs.auth/unauthorized)
+                 ::authority/invalid-json
+                 (do
+                   (log/warn "Could not render authority for OIDC, json invalid")
+                   :com.yetanalytics.lrs.auth/unauthorized)
+                 (throw ex)))))
+         ;; no valid scopes, can't do anything
+         :com.yetanalytics.lrs.auth/unauthorized)})))
+
+;; Authority
+
+(s/fdef resolve-authority-claims
+  :args (s/cat :claims ::oidc/claims)
+  :ret ::oidc/authority-claims)
+
+(defn resolve-authority-claims
```

Comment:
I'm on the fence on whether `get-configuration` or `resolve-authority-claims` should be private functions or not. On one hand they're only used by other functions in this namespace, but I _could_ see them called from other namespaces in the future. (And they certainly need to be public for tests!)